Create delayed delivery table per endpoint by default

SeanFeldman opened the pull request:
Connects to #235

Fixes #237 

Assumptions about implementation laid out in #237.

SeanFeldman commented:
@Particular/azure-maintainers please review

SeanFeldman commented:
@bording could you please review this PR as well?

danielmarbach commented on src/AcceptanceTests/Sending/When_delaying_messages_natively.cs:
```diff
@@ -39,16 +39,16 @@ public class When_delaying_messages_natively : NServiceBusAcceptanceTest
                     sendOptions.DelayDeliveryWith(delay);
-                    c.SW = new Stopwatch();
-                    c.SW.Start();
+                    c.stopwatch = new Stopwatch();
```
Stopwatch.StartNew?

danielmarbach commented on src/AcceptanceTests/Sending/When_delaying_messages_natively.cs:
```diff
@@ -80,25 +80,25 @@ public class When_delaying_messages_natively : NServiceBusAcceptanceTest
         {
             public bool WasCalled { get; set; }
-            public Stopwatch SW { get; set; }
+            public Stopwatch stopwatch { get; set; }
```
Shouldn't this property start with capital `S`?

danielmarbach commented on src/Tests/NServiceBus.Azure.Transports.WindowsAzureStorageQueues.approved.cs:
```diff
@@ -39,6 +39,7 @@ public class DelayedDeliverySettings
     {
         public DelayedDeliverySettings() { }
         public void DisableTimeoutManager() { }
+        public void UseTableName(string delayedMessagesTableName) { }
```
Since tableName is now optional I like this

danielmarbach commented on src/Transport/AzureStorageQueueInfrastructure.cs:
```diff
@@ -33,8 +39,21 @@ internal AzureStorageQueueInfrastructure(SettingsHolder settings, string connect
+        static string GetDelayedDeliveryTableName(string endpointName)
+        {
+            byte[] hashedName;
+            using (var sha1 = new SHA1Managed())
```
I remember a guy who said we should not be using crypto hashes for non crypto stuff ;)

SeanFeldman replied:
Following what this transport is already doing - MD5 for legacy, SHA1 for new version.
Oh, and that guy was Clemens, not Szymon 😜

danielmarbach commented on src/Transport/Config/DelayedDeliverySettings.cs:
```diff
@@ -3,33 +3,30 @@ namespace NServiceBus
 
             const string tableNameRegex = "^[A-Za-z][A-Za-z0-9]{2,62}$";
-            if (new Regex(tableNameRegex).IsMatch(timeoutTableName) == false)
+            if (new Regex(tableNameRegex).IsMatch(delayedMessagesTableName) == false)
```
would it speed up startup if we cache as static?

danielmarbach commented on src/Transport/QueueAddressGenerator.cs:
```diff
@@ -34,7 +34,7 @@ string ShortenQueueNameIfNecessary(string address, string queueName)
             }
 
-            var input = address.Replace('.', '-').ToLowerInvariant(); // this string was used in the past to calculate guid, should stay backward compat
+            var input = address.Replace('.', '-').ToLowerInvariant(); // this string was used in the past to calculate Guid, should stay backward compatibility
```
compatible instead of compatibility?

SeanFeldman replied:
Small English from book 😃

SeanFeldman commented:
@danielmarbach done

danielmarbach commented:
@bording can you have a look at this, when you give your 👍 I will merge

bording commented on src/Transport/AzureStorageQueueInfrastructure.cs:
```diff
@@ -172,34 +191,24 @@ public override string ToTransportAddress(LogicalAddress logicalAddress)
-            await poller.Start(settings, nativeTimeoutsCancellationSource.Token).ConfigureAwait(false);
+            poller = new DelayedMessagesPoller(connectionString, BuildDispatcher(), delayedDeliverySettings.TableName, new BackoffStrategy(maximumWaitTime, peekInterval));
+            nativeDelayedMessagesCancellationSource = new CancellationTokenSource();
+            await poller.Start(settings, nativeDelayedMessagesCancellationSource.Token).ConfigureAwait(false);
```
Something that seems a little weird is that both `DelayedMessagesPoller.Start` and `NativeDelayDelivery.Init` end up calling `CloudTable.CreateIfNotExistsAsync`, and both of these methods are called right here. Seems like it's a bit redundant.

SeanFeldman replied:
Yeah, redundant, but won't cause any potential issues as it will be a no-op if table is already there.

bording replied:
I didn't expect it to cause a problem, just seemed inefficient to be doing that in more than one place. I'll leave it up to you to decide if you want to change it or leave it, though :smile:.

If you are OK with leaving it as-is, then I'd say this is ready to be merged.

SeanFeldman replied:
@bording I'll need to review in VS to know for sure. But after a quick look at it, `NativeDelayDelivery.Init()` could be entirely dropped.

SeanFeldman replied:
Now that I look at it, it's not needed at all. When `DelayedMessagesPoller.Start()` is invoked (before `DelayedDelivery.Init()` is called), it will will ensure that table exists.

SeanFeldman commented:
@danielmarbach @bording I have added additional changes on top in addition to removal of the duplicated table existence check.

Note that the async version of check API seems to have an issue - it returns before table is there, causing tests to fail. Switched to the synchronous version of the API instead with a comment.
